### PR TITLE
impl DispatchFromDyn for Cell and UnsafeCell

### PR DESCRIPTION
**Pull request by dimpolo**
After some fruitful discussion on [Internals](https://internals.rust-lang.org/t/impl-dispatchfromdyn-for-cell-2/16520) here's my first PR to rust-lang/rust 🎉

Please let me know if there's something I missed.

This adds `DispatchFromDyn` impls for `Cell`, `UnsafeCell` and `SyncUnsafeCell`.
An existing test is also expanded to test the `Cell` impl (which requires the `UnsafeCell` impl)

The different `RefCell` types can not implement `DispatchFromDyn` since they have more than one (non ZST) field.

&nbsp; 

**Edit:**
### What:
These changes allow one to make types like `MyRc`(code below), to be object safe method receivers after implementing `DispatchFromDyn` and `Deref` for them.

This allows for code like this:
```rust
struct MyRc<T: ?Sized>(Cell<NonNull<RcBox<T>>>);

/* impls for DispatchFromDyn, CoerceUnsized and Deref for MyRc*/

trait Trait {
    fn foo(self: MyRc<Self>);
}

let impls_trait = ...;
let rc = MyRc::new(impls_trait) as MyRc<dyn Trait>;
rc.foo();
```

Note: `Cell` and `UnsafeCell` won't directly become valid method receivers since they don't implement `Deref`. Making use of these changes requires a wrapper type and nightly features.

### Why:
A custom pointer type with interior mutability allows one to store extra information in the pointer itself.
These changes allow for such a type to be a method receiver.

### Examples:
My use case is a cycle aware custom `Rc` implementation that when dropping a cycle marks some references dangling.

On the [forum](https://internals.rust-lang.org/t/impl-dispatchfromdyn-for-cell/14762/8) andersk mentioned that they track if a `Gc` reference is rooted with an extra bit in the reference itself.

### CODE REVIEW

**Comment by thomcc** [comment]
r? rust-lang/libs-api @rustbot label +T-libs-api -T-libs

**Comment by thomcc** [comment]
> After some fruitful discussion on [Internals](https://internals.rust-lang.org/t/impl-dispatchfromdyn-for-cell-2/16520)

Is this the right link? There are no comments on it.

**Comment by dimpolo** [comment]
Sorry, I should learn that my humour doesn't translate well into text..

Nobody commented, so I thought I'll give this one a try myself :)

**Comment by dimpolo** [comment]
Would it be fair to ping?

The changes should be uncontroversial hopefully. 

Implementing `DispatchFromDyn` enables a type to be used as a `self` type, if that type also implements `Deref`. Since `Cell` and `UnsafeCell` don't implement `Deref` actually making use of these changes requires a custom wrapper type and nighty features (as is shown in the test).

**Comment by m-ou-se** [comment]
@rfcbot merge

**Comment by m-ou-se** [comment]
@rfcbot abort

**Comment by m-ou-se** [comment]
@rfcbot merge

**Comment by m-ou-se** [comment]
Did I break the bot? (:

**Comment by m-ou-se** [comment]
@rfcbot merge

**Comment by m-ou-se** [comment]
The bot should be [fixed](https://rust-lang.zulipchat.com/#narrow/stream/242791-t-infra/topic/I.20broke.20rfcbot/near/286786368) now.

@rfcbot merge

**Comment by BurntSushi** [comment]
> Implementing `DispatchFromDyn` enables a type to be used as a `self` type

Does this mean we should rope the @rust-lang/lang team in on this? I know this is technically adding trait impls and that's libs-api, but this bit makes it sound like the trait impls are actually just a mechanism for a language hook. Increasing the set of types that can be `Self` types seems like a lang change?

**Comment by dimpolo** [comment]
@BurntSushi 
I can't comment about responsibilities. 

About the "set of types that can be `Self` types":
This set seems to be well defined by the requirements of [DispatchFromDyn](https://doc.rust-lang.org/std/ops/trait.DispatchFromDyn.html)
This PR does, however extend this set with types containing interior mutability.

**Comment by dimpolo** [comment]
> Does this mean we should rope the @rust-lang/lang team in on this?

What's the next step here? I'm not familiar with the processes, is there anything I could do to speed things up?

**Comment by nikomatsakis** [comment]
We discussed in our @rust-lang/lang meeting last week. The conclusion was that we'd like to enable experimentation, but we also would like to see a clear comment explaining why the impl was added and what it enables. Right now, it would be very unclear to a reader what is going on, and perhaps appear as if the intention were to enable `Cell<Self>` or some such thing.

**Comment by nikomatsakis** [comment]
Can you re-nominate once this comment is available?

**Comment by dimpolo** [comment]
@nikomatsakis I updated the top level comment with some explanations. If there's anything missing please let me know :)

**Comment by nikomatsakis** [comment]
@dimpolo argh! sorry for the misunderstanding, but what I meant was that we need *comments in the code* explaining why those impls are being added (and in particular that they're meant to enable downstream experimentation).

**Comment by nikomatsakis** [comment]
That said, the changes you put in the OP seem good!

**Comment by dimpolo** [comment]
@nikomatsakis All good
I added comments to the code

File: library/core/src/cell.rs
Comment (Mark-Simulacrum) [inline library/core/src/cell.rs:572]: This probably ought to point at an actual issue, rather than none.

File: library/core/src/cell.rs
Comment (Mark-Simulacrum) [inline library/core/src/cell.rs:2123]: Looks like a stray line here?

**Comment by nikomatsakis** [comment]
@rustbot labels -I-lang-nominated

We discussed this in our meeting today. We are happy with the documentation before merging, we'd like to see a compile-error test showing that using `Cell` (and `CellPtr`) pattern on stable does not compile, since feature gates on impls are a bit subtle.

(Our belief is that this PR does not change stable surface area of Rust, because any practical use of it would require arbitrary-self-types feature, does that sound right @dimpolo ?)

File: library/core/src/cell.rs
Comment (dimpolo) [reply library/core/src/cell.rs:572]: The trait itself currently doesn't have an issue associated with it.
https://github.com/rust-lang/rust/blob/bed4ad65bf7a1cef39e3d66b3670189581b3b073/library/core/src/ops/unsize.rs#L115-L119

We could associate the trait and it's usages with [arbitrary_self_types ](https://github.com/rust-lang/rust/issues/44874) but I'm not sure if that is desired or should be done in this PR.

File: library/core/src/cell.rs
Comment (dimpolo) [reply library/core/src/cell.rs:2123]: I copied this from the above `CoerceUnsized` impl.
I assumed this was meant to signal that there are two unstable features involved here.

**Comment by nikomatsakis** [comment]
@rustbot labels -I-lang-nominated

This test looks good! Removing nomination. Sorry for taking a long time, @dimpolo, our apologies.

File: src/test/ui/feature-gates/feature-gate-dispatch-from-dyn-cell.rs
Comment (tmandry) [inline src/test/ui/feature-gates/feature-gate-dispatch-from-dyn-cell.rs:8]:
```suggestion

trait Trait2 {
    fn cell(self: Cell<Self>); //~ ERROR invalid `self` parameter type: Cell<Self>
}
```

Would this make sense to add?

Comment (dimpolo) [reply src/test/ui/feature-gates/feature-gate-dispatch-from-dyn-cell.rs:8]: If you'd like I'll add.

I don't see what it would be testing though.
I chose `Cell<&Self>` in the test because it could, at least in theory, provide access to `&Self`.
`Cell<Self>` is not able to provide `&Self`.

**Comment by dimpolo** [comment]
Rebased to:
* move tests from `test/` to `tests/`
* update a `.stderr` file

**Comment by dimpolo** [comment]
> This will be merged soon.
Dec 16, 2022

Is something holding up the merge?

File: library/core/src/cell.rs
Comment (dtolnay) [reply library/core/src/cell.rs:2200]: Yeah there has been a longstanding feature request for unstable APIs to be able to be covered by multiple tracking issues / feature gates. For example if there's a `NonZeroI256` type, and an unstable `wrapping_neg` inherent method on it (which is also unstable on a bunch of stable integer types), the method should be covered by gated on _both_ features. We have had a problem in the standard library in the past of accidentally over-stabilizing in similar situations. If `NonZeroI256` were stabilized and its `wrapping_neg` was gated on the `NonZeroI256` feature instead of the `wrapping_neg` feature, it might inadvertently get stabilized during the `NonZeroI256` stabilization without adequate review.

I thought there was a feature request issue filed for this but I wasn't able to find it just now.

The way this is written in the PR is the best we can currently do.

**Review by dtolnay — APPROVED** [review]

**Pull request merged**
